File: ios/RNWortiseSdk/RNWortiseRewarded.swift
import Foundation
import React
import WortiseSDK

@objc(RNWortiseRewarded)
class RNWortiseRewarded: RCTEventEmitter {

    fileprivate static let EVENT_CLICKED        = "onRewardedClicked"
    fileprivate static let EVENT_COMPLETED      = "onRewardedCompleted"
    fileprivate static let EVENT_DISMISSED      = "onRewardedDismissed"
    fileprivate static let EVENT_FAILED_TO_LOAD = "onRewardedFailedToLoad"
    fileprivate static let EVENT_FAILED_TO_SHOW = "onRewardedFailedToShow"
    fileprivate static let EVENT_IMPRESSION     = "onRewardedImpression"
    fileprivate static let EVENT_LOADED         = "onRewardedLoaded"
    fileprivate static let EVENT_REVENUE_PAID   = "onRewardedRevenuePaid"
    fileprivate static let EVENT_SHOWN          = "onRewardedShown"


    fileprivate var rewardedAd: WARewardedAd?


    override static func requiresMainQueueSetup() -> Bool {
        return false
    }


    override func supportedEvents() -> [String]! {
        return [
            RNWortiseRewarded.EVENT_CLICKED,
            RNWortiseRewarded.EVENT_COMPLETED,
            RNWortiseRewarded.EVENT_DISMISSED,
            RNWortiseRewarded.EVENT_FAILED_TO_LOAD,
            RNWortiseRewarded.EVENT_FAILED_TO_SHOW,
            RNWortiseRewarded.EVENT_IMPRESSION,
            RNWortiseRewarded.EVENT_LOADED,
            RNWortiseRewarded.EVENT_REVENUE_PAID,
            RNWortiseRewarded.EVENT_SHOWN
        ]
    }
    

    @objc(cooldownRemainingMs:reject:)
    func cooldownRemainingMs(_ resolve: @escaping RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
        DispatchQueue.main.async { [weak self] in
            resolve(Int((self?.rewardedAd?.cooldownRemaining ?? 0) * 1000))
        }
    }

    @objc
    func destroy() {
        DispatchQueue.main.async { [weak self] in
            self?.rewardedAd?.destroy()
            self?.rewardedAd = nil
        }
    }

    @objc(isAvailable:reject:)
    func isAvailable(_ resolve: @escaping RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
        DispatchQueue.main.async { [weak self] in
            resolve(self?.rewardedAd?.isAvailable ?? false)
        }
    }

    @objc(isInCooldown:reject:)
    func isInCooldown(_ resolve: @escaping RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
        DispatchQueue.main.async { [weak self] in
            resolve(self?.rewardedAd?.isInCooldown ?? false)
        }
    }

    @objc(isShowing:reject:)
    func isShowing(_ resolve: @escaping RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
        DispatchQueue.main.async { [weak self] in
            resolve(self?.rewardedAd?.isShowing ?? false)
        }
    }

    @objc
    func loadAd() {
        DispatchQueue.main.async { [weak self] in
            self?.rewardedAd?.loadAd()
        }
    }

    @objc(setAdUnitId:)
    func setAdUnitId(_ adUnitId: String) {
        DispatchQueue.main.async { [weak self] in

            guard let self = self else {
                return
            }

            self.rewardedAd?.destroy()

            self.rewardedAd = WARewardedAd(adUnitId: adUnitId)
            self.rewardedAd?.delegate = self
        }
    }

    @objc(showAd:reject:)
    func showAd(_ resolve: @escaping RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
        DispatchQueue.main.async { [weak self] in

            guard let rewardedAd = self?.rewardedAd,
                  let controller = RCTPresentedViewController() else {
                resolve(false)
                return
            }

            rewardedAd.showAd(from: controller)

            resolve(true)
        }
    }
}

extension RNWortiseRewarded: WARewardedDelegate {
  
    func didClick(rewardedAd: WARewardedAd) {
        sendEvent(withName: RNWortiseRewarded.EVENT_CLICKED, body: nil)
    }

    func didComplete(rewardedAd: WARewardedAd, reward: WAReward) {
        let body: [String : Any?] = [
            "amount":  reward.amount,
            "label":   reward.label,
            "success": reward.success
        ]

        sendEvent(withName: RNWortiseRewarded.EVENT_COMPLETED, body: body)
    }

    func didDismiss(rewardedAd: WARewardedAd) {
        sendEvent(withName: RNWortiseRewarded.EVENT_DISMISSED, body: nil)
    }

    func didFailToLoad(rewardedAd: WARewardedAd, error: WAAdError) {
        sendEvent(withName: RNWortiseRewarded.EVENT_FAILED_TO_LOAD, body: error.toMap())
    }

    func didFailToShow(rewardedAd: WARewardedAd, error: WAAdError) {
        sendEvent(withName: RNWortiseRewarded.EVENT_FAILED_TO_SHOW, body: error.toMap())
    }

    func didImpress(rewardedAd: WARewardedAd) {
        sendEvent(withName: RNWortiseRewarded.EVENT_IMPRESSION, body: nil)
    }

    func didLoad(rewardedAd: WARewardedAd) {
        sendEvent(withName: RNWortiseRewarded.EVENT_LOADED, body: nil)
    }

    func didPayRevenue(rewardedAd: WARewardedAd, data: WARevenueData) {
        sendEvent(withName: RNWortiseRewarded.EVENT_REVENUE_PAID, body: data.toMap())
    }

    func didShow(rewardedAd: WARewardedAd) {
        sendEvent(withName: RNWortiseRewarded.EVENT_SHOWN, body: nil)
    }
}

File: android/src/main/java/com/wortise/ads/react/RNWortiseRewarded.kt
package com.wortise.ads.react

import android.app.Activity
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.LifecycleEventListener
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.WritableMap
import com.facebook.react.modules.core.DeviceEventManagerModule
import com.wortise.ads.AdError
import com.wortise.ads.RevenueData
import com.wortise.ads.rewarded.RewardedAd
import com.wortise.ads.rewarded.models.Reward
import com.wortise.ads.react.extensions.toWritableMap

class RNWortiseRewarded(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext), RewardedAd.Listener, LifecycleEventListener {

  private val eventEmitter by lazy {
    reactContext.getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
  }

  private var rewardedAd: RewardedAd? = null


  override fun getName(): String = "RNWortiseRewarded"

  @ReactMethod
  fun addListener(eventName: String) {}

  @ReactMethod
  fun destroy() {
    rewardedAd?.destroy()
    rewardedAd = null
  }

  @ReactMethod
  fun isAvailable(promise: Promise) {
    promise.resolve(rewardedAd?.isAvailable == true)
  }

  @ReactMethod
  fun isShowing(promise: Promise) {
    promise.resolve(rewardedAd?.isShowing == true)
  }

  @ReactMethod
  fun loadAd() {
    rewardedAd?.loadAd()
  }

  @ReactMethod
  fun removeListeners(count: Int) {}

  @ReactMethod
  fun setAdUnitId(adUnitId: String) {
    val currentActivity = currentActivity ?: return

    destroy()

    rewardedAd = RewardedAd(currentActivity, adUnitId).also {
      it.listener = this
    }
  }

  @ReactMethod
  fun showAd(promise: Promise) {
    val rewardedAd = rewardedAd

    if (rewardedAd == null) {
      promise.resolve(false)
      return
    }

    currentActivity?.let(rewardedAd::showAd) ?: rewardedAd.showAd()

    promise.resolve(true)
  }


  private fun sendEvent(eventName: String, params: WritableMap? = null) {
    eventEmitter.emit(eventName, params)
  }


  override fun onHostDestroy() {
    destroy()
  }

  override fun onHostPause() {}

  override fun onHostResume() {}

  override fun onRewardedClicked(ad: RewardedAd) {
    sendEvent(EVENT_CLICKED)
  }

  override fun onRewardedCompleted(ad: RewardedAd, reward: Reward) {
    val event = Arguments.createMap()

    event.putInt    ("amount",  reward.amount)
    event.putString ("label",   reward.label)
    event.putBoolean("success", reward.success)

    sendEvent(EVENT_COMPLETED, event)
  }

  override fun onRewardedDismissed(ad: RewardedAd) {
    sendEvent(EVENT_DISMISSED)
  }

  override fun onRewardedFailedToLoad(ad: RewardedAd, error: AdError) {
    sendEvent(EVENT_FAILED_TO_LOAD, error.toWritableMap())
  }

  override fun onRewardedFailedToShow(ad: RewardedAd, error: AdError) {
    sendEvent(EVENT_FAILED_TO_SHOW, error.toWritableMap())
  }

  override fun onRewardedImpression(ad: RewardedAd) {
    sendEvent(EVENT_IMPRESSION)
  }

  override fun onRewardedLoaded(ad: RewardedAd) {
    sendEvent(EVENT_LOADED)
  }

  override fun onRewardedRevenuePaid(ad: RewardedAd, data: RevenueData) {
    sendEvent(EVENT_REVENUE_PAID, data.toWritableMap())
  }

  override fun onRewardedShown(ad: RewardedAd) {
    sendEvent(EVENT_SHOWN)
  }


  companion object {
    const val EVENT_CLICKED        = "onRewardedClicked"
    const val EVENT_COMPLETED      = "onRewardedCompleted"
    const val EVENT_DISMISSED      = "onRewardedDismissed"
    const val EVENT_FAILED_TO_LOAD = "onRewardedFailedToLoad"
    const val EVENT_FAILED_TO_SHOW = "onRewardedFailedToShow"
    const val EVENT_IMPRESSION     = "onRewardedImpression"
    const val EVENT_LOADED         = "onRewardedLoaded"
    const val EVENT_REVENUE_PAID   = "onRewardedRevenuePaid"
    const val EVENT_SHOWN          = "onRewardedShown"
  }
}
